### PR TITLE
feat(messaging): Swift Package Manager support (#13205)

### PR DESCRIPTION
**Pull request by Gustl22**
## Description

SPM Support for firebase_messaging

## Related Issues

#13205

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`).
This will ensure a smooth and quick review process. Updating the `pubspec.yaml` and changelogs is not required.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [ ] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] The analyzer (`melos run analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change.
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/firebase/flutterfire/blob/main/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

**Comment by Gustl22** [comment]
Haven't tested macos though, because one needs a provisioning certificate to run (in this case). Not sure why this is mandatory (but was too lazy honestly to create one, as it should be tested in CI anyways (?))

**Comment by russellwheatley** [comment]
@Gustl22 - if you open the example app for macOS in Xcode, remove the entitlements for **release mode**, you don't need a provisioning profile and the app will build.

File: packages/firebase_messaging/firebase_messaging/ios/firebase_messaging/Package.swift [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+        // Wrapper dependency
+        .product(name: "firebase-core-shared", package: "flutterfire"),
+      ],
+      resources: [
+        .process("Resources"),
+      ],
+      cSettings: [
+        .headerSearchPath("include"),
+        .define("LIBRARY_VERSION", to: "\"\(library_version)\""),
+        .define("LIBRARY_NAME", to: "\"flutter-fire-rc\""),
```
Comment (russellwheatley) [inline packages/firebase_messaging/firebase_messaging/ios/firebase_messaging/Package.swift:116]:
```suggestion
        .define("LIBRARY_NAME", to: "\"flutter-fire-fcm\""),
```

File: packages/firebase_messaging/firebase_messaging/macos/firebase_messaging/Package.swift [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        // Wrapper dependency
+        .product(name: "firebase-core-shared", package: "flutterfire"),
+      ],
+      resources: [
+        .process("Resources"),
+      ],
+      cSettings: [
+        .headerSearchPath("include"),
+        .define("LIBRARY_VERSION", to: "\"\(library_version)\""),
+        .define("LIBRARY_NAME", to: "\"flutter-fire-rc\""),
```
Comment (russellwheatley) [inline packages/firebase_messaging/firebase_messaging/macos/firebase_messaging/Package.swift:118]:
```suggestion
        .define("LIBRARY_NAME", to: "\"flutter-fire-fcm\""),
```

**Review by russellwheatley — CHANGES_REQUESTED** [review]
Looks good so far! A couple of changes and we'll see what the CI says as there might have to be some import updates in ios src/ code

File: packages/firebase_messaging/firebase_messaging/macos/firebase_messaging/Package.swift (hunk 2 above)
Comment (Gustl22) [reply packages/firebase_messaging/firebase_messaging/macos/firebase_messaging/Package.swift]: I honestly didn't know what this identifier stands for ^^

**Comment by Gustl22** [comment]
> if you open the example app for macOS in Xcode, remove the entitlements for release mode, you don't need a provisioning profile and the app will build.

```
flutterfire/packages/firebase_messaging/firebase_messaging/example/macos/Runner.xcodeproj: error: Signing for "Runner" requires a development team. Select a development team in the Signing & Capabilities editor. (in target 'Runner' from project 'Runner')
```

Still not able to build although removing the Signing entitlements in the `Tagets -> Runner -> Build Settings` and setting Signing Team to None in all modes (Debug, Release & Profile).

**Comment by russellwheatley** [comment]
@Gustl22 - I had a quick look at this. It seems flutter_local_notifications package hasn't moved to Swift integration and will therefore run `pod install` which means that we cannot remove the Podfile here: https://github.com/firebase/flutterfire/blob/main/.github/workflows/scripts/swift-integration.dart#L57

You might want to wrap this in an if condition `if(`messaging` == plugin)`.

**Comment by Gustl22** [comment]
Sure, @russellwheatley like that: https://github.com/firebase/flutterfire/pull/16786/commits/8c2f133c6c969230c43a28edcbf9bbbb1a84cfbc ?

**Comment by russellwheatley** [comment]
@Gustl22 - I've made some updates. Should hopefully get this in soon.

**Comment by Gustl22** [comment]
Thank you for taking care of that and investing the time!

**Review by russellwheatley — APPROVED** [review]

**Review by Lyokone — APPROVED** [review]

**Pull request merged**